### DISCUSSION TITLE
Should "Instantiate Model" not produce valid Modelica?

### DISCUSSION THREAD

**Discussion by bilderbuchi** [OpenModelica/OpenModelica#13381] [Q&A] [answered]
I was under the impression that the "Instantiate Model" button in OMEdit produces flattened code that is still valid Modelica. However, I have a model here that when instantiated contains lines with `#` characters, like `  fa := unbox(f(#(u_min)));` or 
```modelica
function Modelica.Math.Nonlinear.solveOneNonlinearEquation "Solve f(u) = 0 in a very reliable and efficient way (f(u_min) and f(u_max) must have different signs)"
  input f<function>(#Real u) => #Real f "Function y = f(u); u is computed so that y=0";
```
It seems my impression is wrong; why does it not?

Background: I am trying to get an MWE for a potential compiler bug by using the test case reducer https://github.com/DRMacIver/shrinkray. This tool mutates a file to reduce its size while preserving the occurrence of the error message, thus getting a smaller MWE exhibiting the same problem.

I want to work on the flattened code to give the reduce more opportunity to reduce across package boundaries etc.
However, when loading the instantiated code, I got (understandably) " Error: Lexer failed to recognize '#(0.0020158'\n""

Could it be that `Modelica.Media.IdealGases.Common.DataRecord` is involved/broken somehow? It seems to be near most of those invalid occurrences of `#`.

Question: How can I flatten a Model to valid Modelica code?

**Comment by perost** [comment, marked as answer]
No, the flat model normally produced by the compiler is not supposed to be valid Modelica, it's mostly just for debugging the frontend. Besides containing internal stuff like boxing/unboxing there are a lot of things that aren't valid Modelica, such as names of functions/models/components being qualified names, components are scalarized so e.g. `Real x[1]` is a scalar and not an array, and so on.

What you're probably looking for is [Base Modelica](https://github.com/modelica/ModelicaSpecification/tree/MCP/0031/RationaleMCP/0031), which you can get by using the `--baseModelica` flag (or the short form `-f`). However, Base Modelica is still being standardized and our implementation is very much still in development, but @casella has had some success using it in some applications.

**Reply by bilderbuchi** [reply]
Thanks! I tried that already, `--baseModelica` also produces invalid code, just with fewer instances of #, in this case of 
```modelica
  function 'Blah.Medium.T_h' "Compute temperature from specific enthalpy"
    input Real 'h'(nominal = 1e5, start = 4.200697462150524e6, max = 1e10, min = -1e10, unit = "J/kg", quantity = "SpecificEnergy");
    output Real 'T'(nominal = 500.0, start = 500.0, max = 6000.0, min = 200.0, displayUnit = "degC", unit = "K", quantity = "ThermodynamicTemperature");
  algorithm
    'T' := 'Modelica.Math.Nonlinear.solveOneNonlinearEquation'(function 'Blah.Medium.T_h.f_nonlinear'(data = #'Modelica.Media.IdealGases.Common.DataRecord'("H2", BOX(0.00201588), BOX(0.0), BOX(4.200697462150524e6), BOX(1000.0), BOX({40783.2321, -800.918604, 8.21470201, -0.01269714457, 1.753605076e-5, -1.20286027e-8, 3.36809349e-12}), BOX({2682.484665, -30.43788844}), BOX({560812.801, -837.150474, 2.975364532, 0.001252249124, -3.74071619e-7, 5.9366252e-11, -3.6069941e-15}), BOX({5339.82441, -2.202774769}), BOX(4124.487568704486)), h = BOX('h')), 200.0, 6000.0, 1e-13);
  end 'Blah.Medium.T_h';
```
(again, in the record)

**Reply by perost** [reply]
Could you share the model you're using?

**Reply by bilderbuchi** [reply]
Sorry, this is internal -- I'm doing this work specifically to get to something shareable (and smaller) :-P 
I'll try to see if I can construct something leading to that situation manually.

**Reply by perost** [reply]
No worries, it should be fixed with #13382. We were marking box/unbox operations in Base Modelica for some reason, but since these are just things we use internally there's little point in adding them to the Base Modelica output.

**Reply by bilderbuchi** [reply]
Ah, great, thank you. I'll test with tomorrow's nightly. :-)

**Reply by bilderbuchi** [reply]
Should I report BaseModelica output that I discover to not be loadable as regular issues? 
I see
```
  function 'Modelica.Math.Nonlinear.solveOneNonlinearEquation' "Solve f(u) = 0 in a very reliable and efficient way (f(u_min) and f(u_max) must have different signs)"
    input f<function>(Real u) => Real y 'f';
    input Real 'u_min';
    input Real 'u_max';
```
where omc (rightfully I think) barfs with "Error: No viable alternative near token: function" at `<function>`.

**Reply by perost** [reply]
Please do.

**Reply by bilderbuchi** [reply]
Done: #13388

**Comment by casella** [comment]
Base Modelica is still an draft proposal. It is a stripped down version of Modelica, without the object orientation. It uses Modelica quoted identifiers to accommodate variable names with dots in them. For technical reasons, it is currently not a strict subset of Modelica, i.e., there are legal Base Modelica models that are not legal Modelica models. But in most cases, you should be able to export a flattened Base Modelica model and then re-load it in OMC without problems.

BTW, for model using arrays, `--baseModelica` flattens the code to scalar variables and equations, but if you also set `--newBackend`, the flat model will be array-preserving. This will be essential in the future when we have full array-preserving toolchains, but it could also be interesting for you now, at least for compactness. If you don't need to change parameters at runtime, you can also add `-d=evaluateAllParameters`, so parameters and derived parameters are evaluated to literal constants at compile time.

@bilderbuchi, if you find any such problems, please report them, we are constantly trying to improve our Base Modelica export feature.

**Reply by bilderbuchi** [reply]
> @bilderbuchi, if you find any such problems, please report them, we are constantly trying to improve our Base Modelica export feature.

I did/do, thanks. :-)

Thanks for the guidance re: the additional flags. Typically I don't have huge arrays, so that is less of a concern. The parameter evaluation might come in handy, though.
